native/src/modules/endpoint/reducers/pushOrRefreshCategory.js

```diff
@@ -0,0 +1,149 @@
+  }
+  const allAvailableLanguages = new Map(category.availableLanguages)
+  allAvailableLanguages.set(language, category.path)
+  return allAvailableLanguages
+}
+
+const extractResultModelsAndChildren = (root: CategoryModel, categoriesMap: CategoriesMapModel, depth: number): {|
+  resultModels: { [path: PathType]: CategoryModel },
+  resultChildren: { [path: PathType]: $ReadOnlyArray<PathType> }
+|} => {
+  /**
+   * Extracts models and children from the (updated) categories map.
+   */
+  const resultModels = {}
+  const resultChildren = {}
+
+  forEachTreeNode(root, (node: CategoryModel) => categoriesMap.getChildren(node), depth, (node, children) => {
+    resultModels[node.path] = node
+    if (children) {
+      resultChildren[node.path] = children.map(child => child.path)
+    }
+  })
+  return {
+    resultModels,
+    resultChildren
+  }
+}
+
+const pushOrRefreshCategory = (state: CityContentStateType,
+  action: PushCategoryActionType | RefreshCategoryActionType): CityContentStateType => {
+  const type = action.type
+  const { categoriesMap, path, depth, key, language, city, resourceCache, cityLanguages } = action.params
+
+  // If there is an error in the old resourceCache, we want to override it
+  const newResourceCache = state.resourceCache.status === 'ready'
+    ? { ...state.resourceCache.value, ...resourceCache }
+    : resourceCache
+
+  const root: ?CategoryModel = categoriesMap.findCategoryByPath(path)
+
+  if (!root) {
+    const route: CategoryRouteStateType = {
+      path: path,
+      depth: depth,
+      language,
+      city,
+      status: 'error',
+      message: `Could not find a category with path '${path}'.`,
+      code: ErrorCodes.PageNotFound
+    }
+
+    return {
+      ...state,
+      categoriesRouteMapping: {
+        ...state.categoriesRouteMapping,
+        [key]: route
+      },
+      resourceCache: {
+        status: 'ready',
+        value: newResourceCache
+      },
+      searchRoute: { categoriesMap }
+    }
+  }
+
+  // Check whether another page in the same city is loading, e.g. because it is being refreshed.
+  // This is important for displaying the loading spinner.
+  const otherPageLoading = Object.values(state.categoriesRouteMapping)
+    // $FlowFixMe Flow does not support Object.values
+    .filter(route => city === route.city && path !== route.path && language === route.language)
+    // $FlowFixMe Flow does not support Object.values
+    .some(route => route.status === 'loading')
+
+  const { resultModels, resultChildren } = extractResultModelsAndChildren(root, categoriesMap, depth)
+
+  // $FlowFixMe Flow can't evaluate the status as it is dynamic
+  const route: CategoryRouteStateType = {
+    path: root.path,
+    models: resultModels,
+    children: resultChildren,
+    depth: depth,
+    allAvailableLanguages: getAllAvailableLanguages(root, city, language, cityLanguages),
+    language,
+    city,
+    status: (otherPageLoading && type === 'PUSH_CATEGORY') ? 'loading' : 'ready'
```

Comment:
In case of `REFRESH_CATEGORY` there is only `ready`, as all open routes are updated with the new content and then considered 'ready'. Meaning, in this case it's irrelevant if another page is loading or not.